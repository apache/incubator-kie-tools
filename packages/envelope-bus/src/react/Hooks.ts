/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ApiDefinition, MessageBusClient, NotificationPropertyNames, SubscriptionCallback } from "../api";
import { useEffect } from "react";

export function useSubscription<Api extends ApiDefinition<Api>, M extends NotificationPropertyNames<Api>>(
  bus: MessageBusClient<Api>,
  method: M,
  callback: SubscriptionCallback<Api, M>
) {
  useEffect(() => {
    const subscription = bus.subscribe(method, callback);
    return () => {
      bus.unsubscribe(method, subscription);
    };
  }, [bus, method, callback]);
}

export function useSubscriptionOnce<Api extends ApiDefinition<Api>, M extends NotificationPropertyNames<Api>>(
  bus: MessageBusClient<Api>,
  method: M,
  callback: SubscriptionCallback<Api, M>
) {
  useEffect(() => {
    let unsubscribed = false;

    const subscription = bus.subscribe(method, (...args) => {
      callback(...args);
      unsubscribed = true;
      bus.unsubscribe(method, subscription);
    });

    return () => {
      if (!unsubscribed) {
        bus.unsubscribe(method, subscription);
      }
    };
  }, [bus, method, callback]);
}
