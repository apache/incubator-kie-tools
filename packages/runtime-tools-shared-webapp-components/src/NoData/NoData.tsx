/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, { useState, useMemo } from "react";
import {
  EmptyState,
  EmptyStateIcon,
  EmptyStateVariant,
  EmptyStateBody,
  EmptyStateHeader,
  EmptyStateFooter,
} from "@patternfly/react-core/dist/js/components/EmptyState";
import { Button } from "@patternfly/react-core/dist/js/components/Button";

import { PageSection } from "@patternfly/react-core/dist/js/components/Page";
import { Bullseye } from "@patternfly/react-core/dist/js/layouts/Bullseye";
import { SearchIcon } from "@patternfly/react-icons/dist/js/icons/search-icon";
import { Navigate, useLocation } from "react-router-dom";
import { OUIAProps, componentOuiaProps } from "@kie-tools/runtime-tools-components/dist/ouiaTools";

export interface IOwnProps {
  defaultPath: string;
  defaultButton: string;
}

export const NoData: React.FC<IOwnProps & OUIAProps> = ({ ouiaId, ouiaSafe, ...props }) => {
  const location = useLocation();

  const prevPath = useMemo(() => {
    return (location.state?.prev ?? props.defaultPath).split("/").filter((item: string) => item);
  }, [location.state, props.defaultPath]);

  const [isRedirect, setIsRedirect] = useState(false);
  const redirectHandler = () => {
    setIsRedirect(true);
  };
  return (
    <>
      {isRedirect && <Navigate replace to={`/${prevPath?.[0] ?? ""}`} />}
      <PageSection isFilled={true} {...componentOuiaProps(ouiaId, "no-data", ouiaSafe ? ouiaSafe : !isRedirect)}>
        <Bullseye>
          <EmptyState variant={EmptyStateVariant.full}>
            <EmptyStateHeader
              titleText={<>{location.state ? location.state.title : "No matches"}</>}
              icon={<EmptyStateIcon icon={SearchIcon} />}
              headingLevel="h1"
            />
            <EmptyStateBody>{location.state ? location.state.description : "No data to display"}</EmptyStateBody>
            <EmptyStateFooter>
              <Button variant="primary" onClick={redirectHandler} data-testid="redirect-button">
                {location.state ? location.state.buttonText : props.defaultButton}
              </Button>
            </EmptyStateFooter>
          </EmptyState>
        </Bullseye>
      </PageSection>
    </>
  );
};
