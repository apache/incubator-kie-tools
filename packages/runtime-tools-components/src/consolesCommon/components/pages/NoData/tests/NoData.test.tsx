/*
 * Copyright 2021 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { NoData } from "../NoData";
import { BrowserRouter as Router } from "react-router-dom";

const props1 = {
  defaultPath: "/processInstances",
  defaultButton: "",
  location: {},
};

const props2 = {
  defaultPath: "/processInstances",
  defaultButton: "",
  location: {
    state: {
      title: "NoData",
      prev: "/processInstances",
      description: "some description",
      buttonText: "button",
    },
  },
};

describe("NoData component tests", () => {
  it("snapshot tests with location object", () => {
    const { container } = render(<NoData {...props1} />);
    expect(container).toMatchSnapshot();
  });

  it("snapshot tests without location object", () => {
    const { container } = render(<NoData {...props2} />);
    expect(container).toMatchSnapshot();
  });
  /* tslint:disable */
  it("redirect button click", async () => {
    const { container } = render(
      <Router>
        <NoData {...props2} />
      </Router>
    );
    const button = screen.getByTestId("redirect-button");
    fireEvent.click(button);
    expect(container).toMatchSnapshot();
  });
});
