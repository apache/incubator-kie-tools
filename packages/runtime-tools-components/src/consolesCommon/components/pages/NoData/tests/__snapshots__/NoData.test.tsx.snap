// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoData component tests redirect button click 1`] = `
<div>
  <section
    class="pf-c-page__main-section pf-m-fill"
    data-ouia-component-type="no-data"
    data-ouia-safe="false"
  >
    <div
      class="pf-l-bullseye"
    >
      <div
        class="pf-c-empty-state"
      >
        <div
          class="pf-c-empty-state__content"
        >
          <svg
            aria-hidden="true"
            class="pf-c-empty-state__icon"
            fill="currentColor"
            height="1em"
            role="img"
            style="vertical-align: -0.125em;"
            viewBox="0 0 512 512"
            width="1em"
          >
            <path
              d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            />
          </svg>
          <h1
            class="pf-c-title pf-m-4xl"
            data-ouia-component-id="OUIA-Generated-Title-3"
            data-ouia-component-type="PF4/Title"
            data-ouia-safe="true"
          >
            NoData
          </h1>
          <div
            class="pf-c-empty-state__body"
          >
            some description
          </div>
          <button
            aria-disabled="false"
            class="pf-c-button pf-m-primary"
            data-ouia-component-id="OUIA-Generated-Button-primary-3"
            data-ouia-component-type="PF4/Button"
            data-ouia-safe="true"
            data-testid="redirect-button"
            type="button"
          >
            button
          </button>
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`NoData component tests snapshot tests with location object 1`] = `
<div>
  <section
    class="pf-c-page__main-section pf-m-fill"
    data-ouia-component-type="no-data"
    data-ouia-safe="true"
  >
    <div
      class="pf-l-bullseye"
    >
      <div
        class="pf-c-empty-state"
      >
        <div
          class="pf-c-empty-state__content"
        >
          <svg
            aria-hidden="true"
            class="pf-c-empty-state__icon"
            fill="currentColor"
            height="1em"
            role="img"
            style="vertical-align: -0.125em;"
            viewBox="0 0 512 512"
            width="1em"
          >
            <path
              d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            />
          </svg>
          <h1
            class="pf-c-title pf-m-4xl"
            data-ouia-component-id="OUIA-Generated-Title-1"
            data-ouia-component-type="PF4/Title"
            data-ouia-safe="true"
          >
            No matches
          </h1>
          <div
            class="pf-c-empty-state__body"
          >
            No data to display
          </div>
          <button
            aria-disabled="false"
            class="pf-c-button pf-m-primary"
            data-ouia-component-id="OUIA-Generated-Button-primary-1"
            data-ouia-component-type="PF4/Button"
            data-ouia-safe="true"
            data-testid="redirect-button"
            type="button"
          >
            
          </button>
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`NoData component tests snapshot tests without location object 1`] = `
<div>
  <section
    class="pf-c-page__main-section pf-m-fill"
    data-ouia-component-type="no-data"
    data-ouia-safe="true"
  >
    <div
      class="pf-l-bullseye"
    >
      <div
        class="pf-c-empty-state"
      >
        <div
          class="pf-c-empty-state__content"
        >
          <svg
            aria-hidden="true"
            class="pf-c-empty-state__icon"
            fill="currentColor"
            height="1em"
            role="img"
            style="vertical-align: -0.125em;"
            viewBox="0 0 512 512"
            width="1em"
          >
            <path
              d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
            />
          </svg>
          <h1
            class="pf-c-title pf-m-4xl"
            data-ouia-component-id="OUIA-Generated-Title-2"
            data-ouia-component-type="PF4/Title"
            data-ouia-safe="true"
          >
            NoData
          </h1>
          <div
            class="pf-c-empty-state__body"
          >
            some description
          </div>
          <button
            aria-disabled="false"
            class="pf-c-button pf-m-primary"
            data-ouia-component-id="OUIA-Generated-Button-primary-2"
            data-ouia-component-type="PF4/Button"
            data-ouia-safe="true"
            data-testid="redirect-button"
            type="button"
          >
            button
          </button>
        </div>
      </div>
    </div>
  </section>
</div>
`;
